R/excel_time_to_numeric.R

```diff
@@ -0,0 +1,166 @@
+#'     \item{A character string that looks like a time.  Choices are 12-hour time ""}
+#'   }
+#' }
+#'
+#' @param time_value A vector of values to convert (see Details)
+#' @param round_seconds Should the output number of seconds be rounded to an
+#'   integer?
+#' @return A vector of numbers >= 0 and <86400
+#' @family Date-time cleaning
+#' @seealso \code{\link{excel_numeric_to_date}}
+#' @export
+excel_time_to_numeric <- function(time_value, round_seconds = TRUE) {
+  UseMethod("excel_time_to_numeric")
+}
+
+#' @export
+excel_time_to_numeric.logical <- function(time_value, round_seconds = TRUE) {
+  if (all(is.na(time_value))) {
+    rep(NA_real_, length(time_value))
+  } else {
+    stop("If given as a logical vector, all values must be ")
+  }
+}
+
+#' @export
+excel_time_to_numeric.numeric <- function(time_value, round_seconds = TRUE) {
+  if (all(is.na(time_value) |
+    time_value >= 0 &
+      time_value < 1)) {
+    seconds <- time_value * 86400
+    if (round_seconds) {
+      seconds <- round(seconds)
+    }
+  } else {
+    stop("When numeric, all `time_value`s must be between 0 and 1 (exclusive of 1)")
+  }
+  seconds
+}
+
+#' @export
+excel_time_to_numeric.POSIXct <- function(time_value, round_seconds = TRUE) {
+  # using trunc removes timezone inconsistency.  Timezones aren't used in Excel.
+  seconds <- as.numeric(time_value) - as.numeric(trunc(time_value, units = "days"))
+  mask_good_seconds <- is.na(seconds) | (seconds >= 0 & seconds < 86400)
+  if (all(mask_good_seconds)) {
+    if (round_seconds) {
+      seconds <- round(seconds)
+    }
+  } else {
+    # This should be impossible except for leap seconds
+    stop(sum(!mask_good_seconds), " `time_value`s were not at or above 0 and below 86400.") # nocov
+  }
+  seconds
+}
+
+#' @export
+excel_time_to_numeric.POSIXlt <- function(time_value, round_seconds = TRUE) {
+  excel_time_to_numeric.POSIXct(
+    as.POSIXct(time_value),
+    round_seconds = round_seconds
+  )
+}
+
+#' @export
+excel_time_to_numeric.character <- function(time_value, round_seconds = TRUE) {
+  ret <- rep(NA_real_, length(time_value))
+  patterns <-
+    list(
+      number = "^0(\\.[0-9]*)?$",
+      # SI numbers have to have the form [number]E-[number] becasue the number
+      # has to be between 0 and 1 and can't be bigger than 1.
+      si_number = "^[1-9](\\.[0-9]*)?E-[0-9]+$",
+      "12hr" = "^([0]?[1-9]|1[0-2]):([0-5][0-9])(?::([0-5][0-9]))? ?([AP]M)$",
+      "24hr" = "^([0-1]?[0-9]|2[0-3]):([0-5][0-9])(?::([0-5][0-9]))?$",
+      # The ".*?" at the end of POSIX is to allow for a time zone, but it allows
+      #   for imperfect parsing if there were just a date and a space.
+      # The the entire time is optional to allow for midnight which shows as
+      #   just the date and time zone.
+      POSIX = "1899-12-31 (?:([0-1]?[0-9]|2[0-3]):([0-5][0-9])(?::([0-5][0-9]))?)?.*?$"
```

Comment:
Will it still match midnight with the trailing space after the 1899-12-31?

Reply:
I added this test which I think confirms that the answer is yes:

```
expect_equal(excel_time_to_numeric(as.POSIXct("1899-12-31 00:01")), 60)
```